include/envoy/http/original_ip_detection.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/address.h"
+
+namespace Envoy {
+namespace Http {
+
+struct OriginalIPDetectionParams {
+  Http::RequestHeaderMap& request_headers;
+};
+
+/**
+ * Interface class for original IP detection extensions.
+ */
+class OriginalIPDetection {
+public:
+  virtual ~OriginalIPDetection() = default;
+
+  /**
+   * Detect the final remote address.
+   *
+   * @param param supplies the OriginalIPDetectionParams params for detection.
+   */
+  virtual Network::Address::InstanceConstSharedPtr
+  detect(struct OriginalIPDetectionParams& params) PURE;
+};
+
+using OriginalIPDetectionSharedPtr = std::shared_ptr<OriginalIPDetection>;
+
+/*
+ * A factory for creating original IP detection extensions.
+ */
+class OriginalIPDetectionFactory : public Envoy::Config::TypedFactory {
+public:
+  ~OriginalIPDetectionFactory() override = default;
```

Comment:
nit: default destructor override shouldn't be needed.

Reply:
Hmm, I think we do this everywhere.. Should I follow-up with a clang-tidy rule to detect these (if possible)?

Reply:
If possible would be lovely but if it's non-trivial no worries.